File: r/diamond/diamond.R
diamond <- function(letter) {
  if (letter == "A") return("A")
  letter_index <- match(letter, LETTERS)
  
  make_row <- function(c) {
    if (c == "A")
      paste0(strrep(" ", letter_index - 1), "A", strrep(" ", letter_index - 1))
    else {
      pos <- match(c, LETTERS)
      outer <- strrep(" ", letter_index - pos)
      inner <- strrep(" ", 2 * pos - 3)
      paste0(outer, c, inner, c, outer)
    }
  }

  top_half = lapply(LETTERS[1:letter_index], make_row)
  bottom_half = rev(top_half)
  paste(c(top_half, bottom_half[-1]), collapse="\n")
}

File: csharp/diamond/Diamond.cs
public static class Diamond
{
    public static string Make(char target)
    {
        if (target == 'A') return "A";
        var letterIndex = target - 'A';
        var rows = new string[2 * letterIndex + 1];
        rows[0] = new string(' ', letterIndex) + "A" + new string(' ', letterIndex);
        rows[2 * letterIndex] = rows[0];
        for (int i = 1; i <= letterIndex; i++)
        {
            var outer = new string(' ', letterIndex - i);
            var inner = new string(' ', 2 * i - 1);
            var letter = (char)('A' + i);
            rows[i] = outer + letter + inner + letter + outer;
            rows[2 * letterIndex - i] = rows[i];
        }

        return string.Join("\n", rows);
    }
}
